cabal-install-solver/src/Distribution/Solver/Modular/MessageUtils.hs

```diff
@@ -0,0 +1,57 @@
+-- | Utility functions providing extra context to cabal error messages
+
+module Distribution.Solver.Modular.MessageUtils (
+  allKnownExtensions,
+  cutoffRange,
+  mostSimilarElement,
+  showUnsupportedExtension,
+  showUnsupportedLanguage,
+  withinRange
+) where
+
+import Data.Foldable (minimumBy)
+import Data.Ord (comparing)
+import Distribution.Pretty (prettyShow) -- from Cabal
+import Language.Haskell.Extension
+         ( Extension(..), Language(..), knownLanguages, knownExtensions )
+import Text.EditDistance ( defaultEditCosts, levenshteinDistance )
+import Text.Printf (printf)
+
+showUnsupportedExtension :: Extension -> String
+showUnsupportedExtension (EnableExtension knownExt) = prettyShow knownExt ++ "which is not supported"
+showUnsupportedExtension (DisableExtension knownExt) = prettyShow knownExt ++ "which is not supported"
+showUnsupportedExtension (UnknownExtension extStr) = formatMessage cutoffRange "extension" extStr (mostSimilarElement extStr allKnownExtensions)
+
+showUnsupportedLanguage :: Language -> String
+showUnsupportedLanguage (UnknownLanguage langStr) = formatMessage cutoffRange "language" langStr (mostSimilarElement langStr (show <$> knownLanguages))
+showUnsupportedLanguage knownLanguage = prettyShow knownLanguage ++ "which is not supported"
+
+allKnownExtensions :: [String]
+allKnownExtensions = enabledExtensions ++ disabledExtensions
+  where
+    enabledExtensions = show <$> knownExtensions
+    disabledExtensions = ("No" ++) . show <$> knownExtensions
+
+-- Measure the levenshtein distance between two strings
+distance :: String -> String -> Int
+distance = levenshteinDistance defaultEditCosts
+
+-- Given an `unknownElement` and a list of `elements` return the element
+-- from the list with the closest levenshtein distance to the `unknownElement`
+mostSimilarElement :: String -> [String] -> String
+mostSimilarElement unknownElement elements = fst . minimumBy (comparing snd) . map mapDist $ elements
+  where
+    mapDist element = (element, distance unknownElement element)
+
+-- Cutoff range for giving a suggested spelling
+cutoffRange :: Int
+cutoffRange = 10
+
+formatMessage :: Int -> String -> String -> String -> String
+formatMessage range elementType element suggestion
+  | withinRange range element suggestion = printf "unknown %s %s did you mean %s?" elementType element suggestion
+  | otherwise = printf "unknown %s %s" elementType element
```

Comment:
`printf` isn't really used in `cabal-install` to concate the messages. Use `concat [...]` or `unwords`. There are examples like

```haskell
            die' verbosity $ unwords
              [ "Failed to download", show uri
              , ": SHA256 don't match; expected:", BS8.unpack (Base16.encode expected)
              , "actual:", BS8.unpack (Base16.encode actual)
              ]

```

in the code already.

Reply:
Fixed